src/backend/distributed/utils/foreign_key_relationship.c

```diff
@@ -149,24 +149,39 @@ GetForeignConstraintRelationshipHelper(Oid relationId, bool isReferencing)
 		return NIL;
 	}
 
-	GetConnectedListHelper(relationNode, &foreignNodeList, isReferencing);
+	List *foreignNodeList = NIL;
+	GetConnectedListHelper(relationshipNode, &foreignNodeList, isReferencing);
 
-	/*
-	 * We need only their OIDs, we get back node list to make their visited
-	 * variable to false for using them iteratively.
-	 */
-	ForeignConstraintRelationshipNode *currentNode = NULL;
-	foreach_ptr(currentNode, foreignNodeList)
-	{
-		foreignConstraintList = lappend_oid(foreignConstraintList,
-											currentNode->relationId);
-		currentNode->visited = false;
-	}
+	/* reset visited flags in foreign key graph */
+	SetRelationNodeListNotVisited(foreignNodeList);
 
 	/* set to false separately, since we don't add itself to foreign node list */
-	relationNode->visited = false;
+	relationshipNode->visited = false;
```

Comment:
high-level remark, not for this PR: it seems questionable that we're setting this state in a global hash map (e.g. what if we OOM and get back here later?).

Reply:
true, in addition to those points, I also don't like keeping such flags in a global data structure,